adr/ADR-43.md

```diff
@@ -0,0 +1,87 @@
+# JetStream Per-Message TTL
+
+| Metadata | Value                     |
+|----------|---------------------------|
+| Date     | 2024-07-11                |
+| Author   | @ripienaar                |
+| Status   | Approved                  |
+| Tags     | jetstream, client, server |
+
+## Context and motivation
+
+Streams support a one-size-fits-all approach to message TTL based on the MaxAge setting. This causes any message in the 
+Stream to expire at that age.
+
+There are numerous uses for a per-message version of this limit, some listed below:
+
+ * KV tombstones are a problem in that they forever clog up the buckets with noise, these could have a TTL to make them expire once not useful anymore
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+ * TODO: user use cases
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL as a duration in a header called `Nats-TTL` followed by the duration as 
+nanoseconds, the sever may not support nanosecond level precision and clients can choose to expose TTL as
+seconds or nanoseconds.
```

Comment:
nano seconds is overkill - should be seconds or millis at most.

Reply:
I think we all agree, but all durations are nanos so I think it best to be consistent - even if we never can get that resolution in precision for this feature.